### PR TITLE
Adds Dance Machine

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over TG's dance machine and all its subsqeuent updates.

I don't think this really needs explanation; it's a dance machine; it plays music, strobes, and sends out sparkling light everywehre.

If admins want, they can upload their own songs to it and customize how frequently it flashes in addition to the length of the song.

Currently unavailable, but I may port over TG's Disco Inferno shuttle...

https://i.gyazo.com/129c49fd27c5d439e51128da5c6b3a76.gif

:cl: Fox McCloud
add: Adds in a Dance Machine
/:cl:

### CODE REVIEW

**Comment by taukausanake** [comment]
I am disappointed. I was hoping for DDR. I guess this is okay though...

Looks really memey though. Can't see this being used except for admin abuse if I'm honest

**Comment by Rumiluntti** [comment]
So it's a useless noisemaker just like the shit jukebox we finally got rid of?
AND it turns everyone into spastic breakdancers?
AND it shits out terrible glitter things that go in circles?

Just why.

**Comment by jughu** [comment]
To dance!

**Comment by Ralta** [comment]
If this is an admin-only thing that only occasionally gets used, I'm cool with that. Otherwise, plsgodno

File: code/game/machinery/dance_machine.dm [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+		if("honk")
+			deejay('sound/items/bikehorn.ogg')
+		if("pump")
+			deejay('sound/weapons/shotgunpump.ogg')
+		if("pop")
+			deejay('sound/weapons/gunshot3.ogg')
+		if("saber")
+			deejay('sound/weapons/saberon.ogg')
+		if("harm")
+			deejay('sound/ai/harmalarm.ogg')
```
Comment (Kyep) [inline code/game/machinery/dance_machine.dm]: This causes compile failure on paradise-test:
code\game\machinery\dance_machine.dm:165:error: 'sound/ai/harmalarm.ogg': cannot find file

Comment (Fox-McCloud) [reply code/game/machinery/dance_machine.dm]: This is what happens when you hastily revert content without a second thought in mind.

File: code/__HELPERS/time.dm [hunk 2]
```diff
@@ -69,4 +69,100 @@ proc/isDay(var/month, var/day)
...
 	var/numSeconds = seconds % 60
 	var/numMinutes = (seconds - numSeconds) / 60
-	return "[numMinutes] [numMinutes > 1 ? "minutes" : "minute"] and [numSeconds] seconds."
\ No newline at end of file
+	return "[numMinutes] [numMinutes > 1 ? "minutes" : "minute"] and [numSeconds] seconds."
+
+//Takes a value of time in deciseconds.
+//Returns a text value of that number in hours, minutes, or seconds.
+/proc/DisplayTimeText(time_value)
+	var/second = time_value*0.1
```
Comment (tigercat2000) [inline code/__HELPERS/time.dm]: Spacing around math operators.

File: code/__HELPERS/time.dm [hunk 3]
```diff
@@ -69,4 +69,100 @@ proc/isDay(var/month, var/day)
...
+	var/second_adjusted = null
+	var/second_rounded = FALSE
+	var/minute = null
+	var/hour = null
+	var/day = null
+
+	if(!second)
+		return "0 seconds"
+	if(second >= 60)
+		minute = round_down(second/60)
```
Comment (tigercat2000) [inline code/__HELPERS/time.dm]: Spacing around math operators.

File: code/__HELPERS/time.dm [hunk 4]
```diff
@@ -69,4 +69,100 @@ proc/isDay(var/month, var/day)
...
+		second = round(second - (minute*60), 0.1)
```
Comment (tigercat2000) [inline code/__HELPERS/time.dm]: Spacing around math operators.

File: code/__HELPERS/time.dm [hunk 5]
```diff
@@ -69,4 +69,100 @@ proc/isDay(var/month, var/day)
...
+					second = "[second] seconds"
+				else
+					second = "[round(second, 0.1)] seconds"
+	else
+		second = null
+
+	if(!minute)
+		return "[second]"
+	if(minute >= 60)
+		hour = round_down(minute/60,1)
```
Comment (tigercat2000) [inline code/__HELPERS/time.dm]: Spacing around math operators.

File: code/__HELPERS/time.dm [hunk 6]
```diff
@@ -69,4 +69,100 @@ proc/isDay(var/month, var/day)
...
+		minute = (minute - (hour*60))
```
Comment (tigercat2000) [inline code/__HELPERS/time.dm]: Spacing around math operators.

File: code/__HELPERS/time.dm [hunk 7]
```diff
@@ -69,4 +69,100 @@ proc/isDay(var/month, var/day)
...
+				minute = " and 1 minute"
+			else
+				minute = "1 minute"
+	else
+		minute = null
+
+	if(!hour)
+		return "[minute][second]"
+	if(hour >= 24)
+		day = round_down(hour/24,1)
```
Comment (tigercat2000) [inline code/__HELPERS/time.dm]: Spacing around math operators.

File: code/__HELPERS/time.dm [hunk 8]
```diff
@@ -69,4 +69,100 @@ proc/isDay(var/month, var/day)
...
+		hour = (hour - (day*24))
```
Comment (tigercat2000) [inline code/__HELPERS/time.dm]: Spacing around math operators.

File: code/game/machinery/dance_machine.dm [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+	playsound(src, S, 300, 1)
+
+/obj/machinery/disco/proc/dance_setup()
+	stop = world.time + selection.song_length
+	var/turf/cen = get_turf(src)
+	FOR_DVIEW(var/turf/t, 3, get_turf(src),INVISIBILITY_LIGHTING)
+		if(t.x == cen.x && t.y > cen.y)
+			var/obj/item/device/flashlight/spotlight/L = new /obj/item/device/flashlight/spotlight(t)
+			L.light_color = "red"
+			L.light_power = 30-(get_dist(src,L)*8)
```
Comment (tigercat2000) [inline code/game/machinery/dance_machine.dm]: Spacing around math operators and in proc arguments. - For all of these blocks.

File: code/game/machinery/dance_machine.dm [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+		if(2 to 3)
+			dance3(M)
+		if(4 to 6)
+			dance4(M)
+		if(7 to 9)
+			dance5(M)
+
+/obj/machinery/disco/proc/dance2(mob/living/M)
+	for(var/i = 1, i < 10, i++)
+		for(var/d in list(NORTH,SOUTH,EAST,WEST,EAST,SOUTH,NORTH,SOUTH,EAST,WEST,EAST,SOUTH))
```
Comment (tigercat2000) [inline code/game/machinery/dance_machine.dm]: Spacing after commas.

File: code/game/machinery/dance_machine.dm [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+
+/obj/machinery/disco/proc/dance3(mob/living/M)
+	var/matrix/initial_matrix = matrix(M.transform)
+	for(var/i in 1 to 75)
+		if(!M)
+			return
+		switch(i)
+			if(1 to 15)
+				initial_matrix = matrix(M.transform)
+				initial_matrix.Translate(0,1)
```
Comment (tigercat2000) [inline code/game/machinery/dance_machine.dm]: Spacing after commas for all of these blocks.

File: code/game/machinery/dance_machine.dm [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+			if(WEST)
+				initial_matrix = matrix(M.transform)
+				initial_matrix.Translate(-3,0)
+				animate(M, transform = initial_matrix, time = 1, loop = 0)
+		sleep(1)
+	M.lying_fix()
+
+
+/obj/machinery/disco/proc/dance4(mob/living/M)
+	var/speed = rand(1,3)
```
Comment (tigercat2000) [inline code/game/machinery/dance_machine.dm]: Spacing.

File: code/game/machinery/dance_machine.dm [hunk 13]
```diff
@@ -0,0 +1,479 @@
...
+	if(charge < 35)
+		charge += 1
+	if(world.time < stop && active)
+		var/sound/song_played = sound(selection.song_path)
+
+		for(var/mob/M in range(10,src))
+			if(!(M in rangers))
+				rangers[M] = TRUE
+				M.playsound_local(get_turf(M), null, 100, channel = CHANNEL_JUKEBOX, S = song_played)
+			if(prob(5+(allowed(M)*4)) && M.canmove)
```
Comment (tigercat2000) [inline code/game/machinery/dance_machine.dm]: Spacing around math operators.

File: code/game/machinery/dance_machine.dm [hunk 14]
```diff
@@ -0,0 +1,479 @@
...
+				dance(M)
+		for(var/mob/L in rangers)
+			if(get_dist(src,L) > 10)
```
Comment (tigercat2000) [inline code/game/machinery/dance_machine.dm]: Spacing in proc arguments.

**Comment by tigercat2000** [comment]
:crystal_ball::man_dancing:

**Pull request merged**